src/nvim/highlight_group.c

```diff
@@ -3077,6 +3081,508 @@ matchitem_T *get_match(win_T *wp, int id)
+        || nmatched > 1
+        || shl->rm.endpos[0].col > mincol) {
+      shl->lnum += shl->rm.startpos[0].lnum;
+      break;                            // useful match found
+    }
+
+    // Restore called_emsg for assert_fails().
+    called_emsg = save_called_emsg;
+  }
+}
+
+/// Advance to the match in window "wp" line "lnum" or past it.
+void prepare_search_hl(win_T *wp, match_T *search_hl, linenr_T lnum)
+  FUNC_ATTR_NONNULL_ALL
+{
+  matchitem_T *cur;             // points to the match list
+  match_T *shl;             // points to search_hl or a match
+  bool shl_flag;                // flag to indicate whether search_hl
+                                // has been processed or not
+
+  // When using a multi-line pattern, start searching at the top
+  // of the window or just after a closed fold.
+  // Do this both for search_hl and the match list.
+  cur = wp->w_match_head;
+  shl_flag = false;
+  while (cur != NULL || shl_flag == false) {
+    if (shl_flag == false) {
+      shl = search_hl;
+      shl_flag = true;
+    } else {
+      shl = &cur->hl;  // -V595
+    }
+    if (shl->rm.regprog != NULL
+        && shl->lnum == 0
+        && re_multiline(shl->rm.regprog)) {
+      if (shl->first_lnum == 0) {
+        for (shl->first_lnum = lnum;
+             shl->first_lnum > wp->w_topline;
+             shl->first_lnum--) {
+          if (hasFoldingWin(wp, shl->first_lnum - 1, NULL, NULL, true, NULL)) {
+            break;
+          }
+        }
+      }
+      if (cur != NULL) {
+        cur->pos.cur = 0;
+      }
+      bool pos_inprogress = true;  // mark that a position match search is
+                                   // in progress
+      int n = 0;
+      while (shl->first_lnum < lnum && (shl->rm.regprog != NULL
+                                        || (cur != NULL && pos_inprogress))) {
+        next_search_hl(wp, search_hl, shl, shl->first_lnum, (colnr_T)n,
+                       shl == search_hl ? NULL : cur);
+        pos_inprogress = !(cur == NULL || cur->pos.cur == 0);
+        if (shl->lnum != 0) {
+          shl->first_lnum = shl->lnum
+                            + shl->rm.endpos[0].lnum
+                            - shl->rm.startpos[0].lnum;
+          n = shl->rm.endpos[0].col;
+        } else {
+          shl->first_lnum++;
+          n = 0;
+        }
+      }
+    }
+    if (shl != search_hl && cur != NULL) {
+      cur = cur->next;
+    }
+  }
+}
+
+/// Prepare for 'hlsearch' and match highlighting in one window line.
+/// Return TRUE if there is such highlighting and set "search_attr" to the
+/// current highlight attribute.
+int prepare_search_hl_line(win_T *wp, linenr_T lnum, colnr_T mincol, char_u **line,
+                           match_T *search_hl, int *search_attr, bool *search_attr_from_match,
+                           bool has_fold, bool end_fill)
+{
+  // Handle highlighting the last used search pattern and matches.
+  // Do this for both search_hl and the match list.
```

Comment:
I feel this comment looks a bit strange. Maybe it is better to put this above the `while`.